service/collector/endpoint.go

```diff
@@ -47,23 +61,212 @@ func NewEndpoint(config EndpointConfig) (*Endpoint, error) {
+			continue
+		}
+
+		endpoint, err := e.k8sClient.CoreV1().Endpoints(kvmConfig.Name).Get(workerEndpoint, getOpts)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		ipList := getEndpointIps(endpoint)
+		for _, ip := range ipList {
+			err := e.ingressEndpointUpClassicHttp(ip, ingresSchemeHttp, ingressPortHttp)
+
+			if err == nil {
+				// ingress endpoint check was successful, no proxy-protocol
+				ch <- prometheus.MustNewConstMetric(
+					endpointLabelsDesc,
+					prometheus.GaugeValue,
+					ingressCheckSucesfull,
+					kvmConfig.Name,
+					ip,
+					ingresSchemeHttp,
+					"false",
+				)
+
+			} else {
+				// ingress endpoint check error, but we might just hit proxy protocol so try check with proxy-protocol enabled
+				err := e.ingressEndpointUpProxyProtocol(ip, ingresSchemeHttp, ingressPortHttp)
+
+				if err == nil {
+					// ingress endpoint check was successful, proxy-protocol enabled
+					ch <- prometheus.MustNewConstMetric(
+						endpointLabelsDesc,
+						prometheus.GaugeValue,
+						ingressCheckSucesfull,
+						kvmConfig.Name,
+						ip,
+						ingresSchemeHttp,
+						"true",
+					)
+				}
+				if err != nil {
+					// ingress endpoint check failed
+					// ingress endpoint check was successful, proxy-protocol enabled
+					ch <- prometheus.MustNewConstMetric(
+						endpointLabelsDesc,
+						prometheus.GaugeValue,
+						ingressCheckFailure,
+						kvmConfig.Name,
+						ip,
+						ingresSchemeHttp,
+						"unknown",
+					)
+				}
+			}
+		}
+	}
 
 	return nil
 }
 
-func (i *Endpoint) Describe(ch chan<- *prometheus.Desc) error {
+func (e *Endpoint) Describe(ch chan<- *prometheus.Desc) error {
 	ch <- endpointLabelsDesc
 	return nil
 }
+
+func getEndpointIps(endpoint *corev1.Endpoints) []string {
+	var ipList []string
+	for _, subset := range endpoint.Subsets {
+		for _, address := range subset.Addresses {
+			ipList = append(ipList, address.IP)
+		}
+	}
+
+	return ipList
+}
+
+func (e *Endpoint) buildHttpRequest(ipAddress string, scheme string, port int) (*http.Request, error) {
+	u := url.URL{
+		Host:   fmt.Sprintf("%s:%d", ipAddress, port),
+		Path:   "healthz",
+		Scheme: scheme,
+	}
+
+	// be sure to close idle connection after health check is finished
+	defer e.httpTransport.CloseIdleConnections()
+
+	req, err := http.NewRequest("GET", u.String(), nil)
+	if err != nil {
+		return nil, microerror.Maskf(err, "unable to construct health check request")
+	}
+
+	// close connection after health check request (the TCP connection gets
+	// closed by deferred s.tr.CloseIdleConnections()).
+	req.Header.Add("Connection", "close")
+
+	return req, nil
+}
+
+// ingressEndpointUpClassicHttp send http packet to the endpoint to ensure target ingress endpoint ip is up
+func (e *Endpoint) ingressEndpointUpClassicHttp(ipAddress string, scheme string, port int) error {
+	req, err := e.buildHttpRequest(ipAddress, scheme, port)
+	if err != nil {
+		// failed to build http req
+		return microerror.Mask(err)
+	}
+
```

Comment:
I think idle HTTP connections should be closed here.

Reply:
done